// module/dist/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createModule = createModule;
const node_path_1 = __importDefault(require("node:path"));
const promises_1 = __importDefault(require("node:fs/promises"));
function createModule(dir, system, identifier) {
    return __awaiter(this, void 0, void 0, function* () {
        const name = node_path_1.default.basename(dir);
        const extension = identifier === 'file-ext' ? (system === 'esm' ? 'mjs' : 'cjs') : 'js';
        yield promises_1.default.mkdir(dir, { recursive: true });
        return Promise.all([
            createIndexScript(),
            createCliScript(),
            createPackageJson(),
            createGitignore(),
            createReadme(),
        ]);
        function createIndexScript() {
            return __awaiter(this, void 0, void 0, function* () {
                return createFile(`index.${extension}`, system === 'esm'
                    ? `export default function () {\n  console.log('${name}');\n}\n`
                    : `'use strict';\n\nmodule.exports = function () {\n  console.log('${name}');\n};\n`);
            });
        }
        function createCliScript() {
            return __awaiter(this, void 0, void 0, function* () {
                return createFile(`cli.${extension}`, system === 'esm'
                    ? `#!/usr/bin/env node\n\nimport fn from './';\n\nfn();\n`
                    : `#!/usr/bin/env node\n\n'use strict';\n\nconst fn = require('./');\n\nfn();\n`, 0o755);
            });
        }
        function createPackageJson() {
            return __awaiter(this, void 0, void 0, function* () {
                return createFile('package.json', `${JSON.stringify({
                    private: true,
                    name,
                    version: '0.0.0',
                    type: identifier === 'package-type'
                        ? system === 'esm'
                            ? 'module'
                            : 'commonjs'
                        : undefined,
                    main: `index.${extension}`,
                    bin: `cli.${extension}`,
                    dependencies: {},
                    devDependencies: {},
                }, null, 2)}\n`);
            });
        }
        function createGitignore() {
            return __awaiter(this, void 0, void 0, function* () {
                return createFile('.gitignore', `node_modules/\n`);
            });
        }
        function createReadme() {
            return __awaiter(this, void 0, void 0, function* () {
                return createFile('README.md', `# ${name}\n\nTODO.\n`);
            });
        }
        function createFile(fileName, fileContent, mode) {
            return __awaiter(this, void 0, void 0, function* () {
                const filePath = node_path_1.default.join(dir, fileName);
                yield promises_1.default.writeFile(filePath, fileContent, { mode });
                return filePath;
            });
        }
    });
}
